src/types/utils.cpp

```diff
@@ -631,6 +629,214 @@ bool Utils::IsValidHandle(const HANDLE handle) noexcept
+        /* ObjectAttributes  */ &objectAttributes,
+        /* IoStatusBlock     */ &statusBlock,
+        /* ShareAccess       */ FILE_SHARE_READ | FILE_SHARE_WRITE,
+        /* CreateDisposition */ FILE_CREATE,
+        /* CreateOptions     */ 0, // would be FILE_SYNCHRONOUS_IO_NONALERT for a synchronous pipe
+        /* NamedPipeType     */ FILE_PIPE_BYTE_STREAM_TYPE,
+        /* ReadMode          */ FILE_PIPE_BYTE_STREAM_MODE,
+        /* CompletionMode    */ FILE_PIPE_QUEUE_OPERATION, // would be FILE_PIPE_COMPLETE_OPERATION for PIPE_NOWAIT
+        /* MaximumInstances  */ 1,
+        /* InboundQuota      */ bufferSize,
+        /* OutboundQuota     */ bufferSize,
+        /* DefaultTimeout    */ &timeout));
+
+    switch (openMode)
+    {
+    case PIPE_ACCESS_INBOUND:
+        desiredAccess = SYNCHRONIZE | GENERIC_WRITE | FILE_READ_ATTRIBUTES;
+        shareAccess = FILE_SHARE_READ;
+        break;
+    case PIPE_ACCESS_OUTBOUND:
+        desiredAccess = SYNCHRONIZE | GENERIC_READ | FILE_WRITE_ATTRIBUTES;
+        shareAccess = FILE_SHARE_WRITE;
+        break;
+    default:
+        desiredAccess = SYNCHRONIZE | GENERIC_READ | GENERIC_WRITE;
+        shareAccess = FILE_SHARE_READ | FILE_SHARE_WRITE;
+        break;
+    }
+
+    wil::unique_hfile client;
+    objectAttributes.RootDirectory = server.get();
+    THROW_IF_NTSTATUS_FAILED(NtCreateFile(
+        /* FileHandle        */ client.addressof(),
+        /* DesiredAccess     */ desiredAccess,
+        /* ObjectAttributes  */ &objectAttributes,
+        /* IoStatusBlock     */ &statusBlock,
+        /* AllocationSize    */ nullptr,
+        /* FileAttributes    */ 0,
+        /* ShareAccess       */ shareAccess,
+        /* CreateDisposition */ FILE_OPEN,
+        /* CreateOptions     */ FILE_NON_DIRECTORY_FILE, // would include FILE_SYNCHRONOUS_IO_NONALERT for a synchronous pipe
+        /* EaBuffer          */ nullptr,
+        /* EaLength          */ 0));
+
+    return { std::move(server), std::move(client) };
+}
+
+// GetOverlappedResult() for professionals! Only for single-threaded use.
+//
+// GetOverlappedResult() used to have a neat optimization where it would only call WaitForSingleObject() if the state was STATUS_PENDING.
+// That got removed in Windows 7, because people kept starting a read/write on one thread and called GetOverlappedResult() on another.
+// When the OS sets Internal from STATUS_PENDING to 0 (= done) and then flags the hEvent, that doesn't happen atomically.
+// This results in a race condition if a OVERLAPPED is used across threads.
+HRESULT Utils::GetOverlappedResultSameThread(const OVERLAPPED* overlapped, DWORD* bytesTransferred) noexcept
+{
+    assert(overlapped != nullptr);
+    assert(overlapped->hEvent != nullptr);
+    assert(bytesTransferred != nullptr);
+
+    __assume(overlapped != nullptr);
+    __assume(overlapped->hEvent != nullptr);
+    __assume(bytesTransferred != nullptr);
+
+    if (overlapped->Internal == STATUS_PENDING)

```

Comment:
`Internal` - is this documented anywhere?

Reply:
Yep: https://learn.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-overlapped
It does say "its behavior may change" but I think that's a minor concern for us, given that we're effectively a system component. It also hasn't changed since "Windows NT 1.0" (I honestly don't know but basically since forever). The reason it's unlikely to change is because the two `Internal` fields overlap with NT's `IO_STATUS_BLOCK` struct, whose 2 fields are publicly documented and don't mention this "its behavior may change".